folly/io/IOStreamBuf-inl.h

```diff
@@ -0,0 +1,264 @@
+#include <folly/Likely.h>
+
+namespace folly {
+
+template <typename CharT, typename Traits>
+IOStreamBuf<CharT,Traits>::IOStreamBuf(const folly::IOBuf* head)
+  : std::basic_streambuf<CharT,Traits>(),
+    head_(head),
+    gcur_(head) {
+  csetg(gcur_->data(), gcur_->data(), gcur_->tail());
+}
+
+template <typename CharT, typename Traits>
+void IOStreamBuf<CharT,Traits>::swap(IOStreamBuf<CharT,Traits>& rhs) {
+  std::basic_streambuf<CharT,Traits>::swap(rhs);
+  std::swap(head_, rhs.head_);
+  std::swap(gcur_, rhs.gcur_);
+}
+
+template <typename CharT, typename Traits>
+const typename IOStreamBuf<CharT,Traits>::pos_type
+IOStreamBuf<CharT,Traits>::badoff =
+  static_cast<typename IOStreamBuf<CharT,Traits>::pos_type>(
+    static_cast<typename IOStreamBuf<CharT,Traits>::off_type>(-1));
+
+// This is called either to rewind the get area (because gptr() == eback())
+// or to attempt to put back a non-matching character (which we disallow
+// on non-mutable IOBufs).
+template <typename CharT, typename Traits>
+typename IOStreamBuf<CharT,Traits>::int_type
+IOStreamBuf<CharT,Traits>::pbackfail(int_type c) {
+  if (this->gptr() != this->eback())
+    return traits_type::eof(); // trying to putback non-matching character
+
+  if (gcur_ == head_) {
+    // Already at beginning of first IOBuf
+    return traits_type::eof();
+  }
+
+  // Find the next preceding non-empty IOBuf, back to head_
+  // Ensure the object state is not modified unless an earlier input sequence
+  // can be found.
+  const IOBuf* prev = gcur_;
+  do {
+    prev = prev->prev();
+  } while (prev->length() == 0 && prev != head_);
```

Comment:
Looks like IOBuf::Iterator is a forward-only iterator whereas this code needs to iterate backwards to the head. Furthermore it needs access to the IOBuf object itself but the IOBuf::Iterator only exposes a ByteRange (ie. the data within the IOBuf).